### PR TITLE
odo describe component -o json

### LINKED ISSUE #5660: `odo describe component -o json`
- [ ] without any additional flags, it describes the current component (devfile in the current directory)
- [ ] can be used with `--name` to describe other components existing on the cluster
- [ ]  `--namespace` is optional and can be used only with `--name` flag

non-json output is part of https://github.com/redhat-developer/odo/issues/5661

```
$ odo describe component -o json
```
```json
{
  "devfilePath": "/path/to/devfile.yaml",
  "devfileData": {
    "devfile": <content of devfile>
  },
  "supportedOdoFeatures": {
      "dev": true,
      "debug": true,
      "deploy": true
  },
  "devForwardedPorts": [
  { 
    "containerName": "cont1",
    "localPort": 40001,
    "containerPort": 8080,
    "address": "localhost",
  }
  ],
  "runningIn": ["Dev", "Deploy"], # or just "Dev", "Deploy", "Unknown", or []
  "managedBy": "odo"  
}
```

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind feature

**What does this PR do / why we need it:**

**Which issue(s) this PR fixes:**

Fixes #5660 

**PR acceptance criteria:**

- [x] Unit test 

- [x] Integration test 

- [x] Documentation : JSON output documentation written, the command reference doc will be completed as part of #5661

**How to test changes / Special notes to the reviewer:**

Doc preview:
- https://deploy-preview-5720--odo-docusaurus-preview.netlify.app/docs/3.0.0/command-reference/json-output#odo-describe-component--o-json
- https://deploy-preview-5720--odo-docusaurus-preview.netlify.app/docs/3.0.0/command-reference/describe (to be completed as part of #5661)

### CODE REVIEW

**Review by rm3l — CHANGES_REQUESTED** [review]
Forgot to submit this very first batch of comments :) I am yet to review the other files.

File: pkg/component/component_test.go [hunk 1]
```diff
@@ -272,3 +149,115 @@ func getUnstructured(name, kind, apiVersion, managed, componentType, namespace s
...
+				client: func(ctrl *gomock.Controller) kclient.ClientInterface {
+					c := kclient.NewMockClientInterface(ctrl)
+					c.EXPECT().GetAllResourcesFromSelector(gomock.Any(), gomock.Any()).Return(nil, errors.New("error"))
+					return c
+				},
+				name:      "aname",
+				namespace: "anamespace",
+			},
+			want: []api.RunningMode{api.RunningModeUnknown},
+		},
```
Comment (rm3l) [inline pkg/component/component_test.go:248]: Could we add a case testing an empty list returned by `GetAllResourcesFromSelector`.
At first sight, I thought this was covered by the `No resources` test, but this makes `GetAllResourcesFromSelector` return resources that are ignored.

File: pkg/odo/util/cmdutils.go [hunk 2]
```diff
@@ -20,7 +21,7 @@ func LogErrorAndExit(err error, context string, a ...interface{}) {
 		if log.IsJSON() {
 
 			// Machine readble error output
-			machineOutput := machineoutput.GenericError{
+			machineOutput := api.GenericError{
```
Comment (rm3l) [inline pkg/odo/util/cmdutils.go:24]: I noticed that error messages for JSON output are written to stderr:

```
❯ annotate-output odo describe component --name cmp-should-not-exist --namespace 5660-describe-component-json -o json
11:32:05 I: Started odo describe component --name cmp-should-not-exist --namespace test -o json
11:32:05 E: {
11:32:05 E:     "message": "no component found with name \"cmp-should-not-exist\" in the namespace \"5660-describe-component-json\""
11:32:05 E: }
11:32:05 I: Finished with exitcode 1
```

I think the JSON error message is the result of the operation and should be written to stdout, so it can be piped and used by other tools. WDYT?

Comment (rm3l) [reply pkg/odo/util/cmdutils.go:24]: Same issue if I purposely run a wrong subcommand. Both the help message and the JSON output are written to stderr:

```
❯ annotate-output odo describe anything -o json
11:37:42 I: Started odo describe anything -o json
11:37:42 E: Error: Subcommand not found, use one of the available commands: component
11:37:42 E: Usage:
11:37:42 E:   odo describe [flags]
11:37:42 E:   odo describe [command]
11:37:42 E: 
11:37:42 E: Available Commands:
11:37:42 E:   component   Describe a component
11:37:42 E: 
11:37:42 E: Flags:
11:37:42 E:   -h, --help   Help for describe
11:37:42 E: 
11:37:42 E: Additional Flags:
11:37:42 E:       --kubeconfig string    Paths to a kubeconfig. Only required if out-of-cluster.
11:37:42 E:   -v, --v Level              Number for the log level verbosity. Level varies from 0 to 9 (default 0).
11:37:42 E:       --vmodule moduleSpec   Comma-separated list of pattern=N settings for file-filtered logging
11:37:42 E: 
11:37:42 E: Use "odo describe [command] --help" for more information about a command.
11:37:42 E: 
11:37:42 E: {
11:37:42 E:     "message": "Subcommand not found, use one of the available commands: component"
11:37:42 E: }
11:37:42 I: Finished with exitcode 1
```

**Review by rm3l — CHANGES_REQUESTED** [review]

File: pkg/odo/util/cmdutils.go (hunk 2 above)
Comment (feloy) [reply pkg/odo/util/cmdutils.go:24]: > I noticed that error messages for JSON output are written to stderr:
> 
> ```
> ❯ annotate-output odo describe component --name cmp-should-not-exist --namespace 5660-describe-component-json -o json
> 11:32:05 I: Started odo describe component --name cmp-should-not-exist --namespace test -o json
> 11:32:05 E: {
> 11:32:05 E:     "message": "no component found with name \"cmp-should-not-exist\" in the namespace \"5660-describe-component-json\""
> 11:32:05 E: }
> 11:32:05 I: Finished with exitcode 1
> ```
> 
> This is certainly a wider issue related to error handling for machine-readable output, but I think the JSON error message is the result of the operation and should be written to stdout, so it can be piped and used by other tools. WDYT?

We have documented that in case of error, the message is written to stderr:
https://odo.dev/docs/3.0.0/command-reference/json-output

Let's wait for feedback from plugins developers to fix this if necessary

Comment (rm3l) [reply pkg/odo/util/cmdutils.go:24]: > We have documented that in case of error, the message is written to stderr: https://odo.dev/docs/3.0.0/command-reference/json-output
> 
I see. Great to have this documented. I remember I read this already :-)

> Let's wait for feedback from plugins developers to fix this if necessary
> 
Okay, that makes sense.

File: pkg/component/component_test.go (hunk 1 above)
Comment (feloy) [reply pkg/component/component_test.go:276]: Yes! I think that made me fix a bug for some edge case
Thanks

Comment (rm3l) [reply pkg/component/component_test.go:276]: Indeed. I see the edge case fixed by the filtered list :-D

**Review by rm3l — APPROVED** [review]
